In [1]:
# import the necessary packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix,classification_report

In [2]:
df = pd.read_csv('mnist.csv')       # read the csv file

df_feat = df.drop('label',axis=1)   # drop the label column

label_y = df['label']               # label column

In [3]:
max_acc = np.array([[]])            # array to store the maximum accuracy for each train size

for j in range(1,4):
    j=j/10
    print("Train Size: ",1-j)
    X_train,X_test,Y_train,Y_test = train_test_split(df_feat,label_y,test_size=j,random_state=72) # split the data

    acc = np.array([])              # array to store the accuracy for each k value

    for i in range(1,11):           # loop for k values
        knn = KNeighborsClassifier(n_neighbors=i)   # create the model
        knn.fit(X_train,Y_train)                # fit the model

        pred = knn.predict(X_test)              # predict the values
        print("---------",i,"--------")
        report = classification_report(Y_test,pred,output_dict=True)    # classification report
        print("Accuracy: ",report['accuracy'])                       # accuracy
        acc = np.append(acc,report['accuracy'])
        print("---------------------")

    print("Best Accuracy: ",acc.max().round(3)," k value: ",acc.argmax()+1) # best accuracy and k value
    max_acc=np.append(max_acc,acc.max().round(3))                      # append the best accuracy to the array

print("Best Accuracy: ",max_acc.max()," at Train Size = ",1-(max_acc.argmax()+1)/10)    # best accuracy and train size

Train Size:  0.9
--------- 1 --------
Accuracy:  0.9678571428571429
---------------------
--------- 2 --------
Accuracy:  0.9638095238095238
---------------------
--------- 3 --------
Accuracy:  0.9676190476190476
---------------------
--------- 4 --------
Accuracy:  0.969047619047619
---------------------
--------- 5 --------
Accuracy:  0.9661904761904762
---------------------
--------- 6 --------
Accuracy:  0.9673809523809523
---------------------
--------- 7 --------
Accuracy:  0.9657142857142857
---------------------
--------- 8 --------
Accuracy:  0.9635714285714285
---------------------
--------- 9 --------
Accuracy:  0.9633333333333334
---------------------
--------- 10 --------
Accuracy:  0.9616666666666667
---------------------
Best Accuracy:  0.969  k value:  4
Train Size:  0.8
--------- 1 --------
Accuracy:  0.9654761904761905
---------------------
--------- 2 --------
Accuracy:  0.9604761904761905
---------------------
--------- 3 --------
Accuracy:  0.9673809523809523
----

## Accuracy

We got the highest accuracy of 0.969 at k=4 and train size of 0.9. We have taken k values from 1 to 10 and split the training sample into sizes of 70%, 80%, and 90%. We have used the same test sample for all the cases. This accuracy is comparable to that of Radnom Forest Classifier.